js/data/feature_tree.js

```diff
@@ -88,80 +152,181 @@ FeatureTree.prototype.query = function(args, styleLayersByID) {
         maxY = Math.max(maxY, p.y);
     }
 
-    var bounds = [minX - additionalRadius, minY - additionalRadius, maxX + additionalRadius, maxY + additionalRadius];
-    var symbolQueryBox = new CollisionBox(new Point(minX, minY), 0, 0, maxX - minX, maxY - minY, args.scale, null);
+    var matching = this.grid.query(minX - additionalRadius, minY - additionalRadius, maxX + additionalRadius, maxY + additionalRadius);
+    matching.sort(topDown);
+    var match = this.featureIndexArray.get(0);
+    this.filterMatching(result, matching, match, queryGeometry, filter, params.layers, styleLayersByID, args.bearing, pixelsToTileUnits, returnGeoJSON);
+
+    var matchingSymbols = this.collisionTile.queryRenderedSymbols(minX, minY, maxX, maxY, args.scale);
+    var match2 = this.collisionTile.collisionBoxArray.get(0);
+    matchingSymbols.sort();
+    this.filterMatching(result, matchingSymbols, match2, queryGeometry, filter, params.layers, styleLayersByID, args.bearing, pixelsToTileUnits, returnGeoJSON);
+
+    return result;
+};
+
+function topDown(a, b) {
+    return b - a;
+}
 
-    var matching = this.rtree.search(bounds).concat(this.collisionTile.getFeaturesAt(symbolQueryBox, args.scale));
+function getLineWidth(paint) {
+    if (paint['line-gap-width'] > 0) {
+        return paint['line-gap-width'] + 2 * paint['line-width'];
+    } else {
+        return paint['line-width'];
+    }
+}
 
+FeatureTree.prototype.filterMatching = function(result, matching, match, queryGeometry, filter, filterLayerIDs, styleLayersByID, bearing, pixelsToTileUnits, returnGeoJSON) {
+    var seen = {};
     for (var k = 0; k < matching.length; k++) {
-        var feature = matching[k].feature,
-            layerIDs = matching[k].layerIDs;
+        var index = matching[k];
 
-        if (!filter(feature)) continue;
+        if (seen[index]) continue;
+        seen[index] = true;
```

Comment:
Ah! I understand now. I was just asking for my own understanding. 

If you think that diff :point_up_2: will work in all cases (incl GeoJSON) then let's do it.